### PR TITLE
Topic recognition - OASIS brain data segmentation using UNet

### PR DESCRIPTION
**Pull request by Calvin-Shaji**
Tasks done ->
1.) Train UNet model on training data set.
2.) Evaluate the model on the validation data set.
3.) Generate predictions for the test data set.

Dice similarity coefficient of >0.9 achieved.

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
- Sufficiently detailed commit messages
- You are supposed to use *improved* U-Net
- Wrong target branch for PR
- Code do not appear to be valid. Output class of 1 is incorrect for the data. Thus reported results are invalid.
- No qualitative results to show the model at work.
- incorrect instructions in `README.MD`

**Comment by Calvin-Shaji** [comment]
- README.md is placed inside recognition/calvin directory.
- The code is generating images as results for the test data. It is generating valid images inside the results folder. Could you please explain why it's invalid since the output images are being produced?
- The specs were slightly confusing on the pull request aspect. My interpretation was that I needed to create a pull request from the topic-recognition branch of my repository to the master branch of the original repository. Am I allowed to make a new pull request to the topic-recognition branch now?

**Comment by SiyuLiu0329** [comment]
> * README.md is placed inside recognition/calvin directory.
> * The code is generating images as results for the test data. It is generating valid images inside the results folder. Could you please explain why it's invalid since the output images are being produced?
> * The specs were slightly confusing on the pull request aspect. My interpretation was that I needed to create a pull request from the topic-recognition branch of my repository to the master branch of the original repository. Am I allowed to make a new pull request to the topic-recognition branch now?

`conv10 = Conv2D(filters=1, kernel_size=1, padding='valid', activation='sigmoid', kernel_initializer='glorot_uniform')(conv9)` would not have worked since OASIS is a 4-class segmentation problem.
Where is the results directory you mentioned?

**Comment by Calvin-Shaji** [comment]
I have added the results directory. It didn't get uploaded earlier because it was in the same folder as the provided files, and that had been included in gitignore.

**Comment by SiyuLiu0329** [comment]
> I have added the results directory. It didn't get uploaded earlier because it was in the same folder as the provided files, and that had been included in gitignore.

Results look okay, you still need to clearify how the results came to be using the code you submitted at the time. How did you get the classification results using 1-class output + sigmoid?

**Comment by Calvin-Shaji** [comment]
The dice coefficient gave a value of over 0.9
I had included some of the images of the result in the turn it in pdf where you can see the result is over 0.9

**Comment by SiyuLiu0329** [comment]
> The dice coefficient gave a value of over 0.9
> I had included some of the images of the result in the turn it in pdf where you can see the result is over 0.9

The results need to be justified. We are required to question any implementation details we are not sure of.

**Comment by Calvin-Shaji** [comment]
- The output is being produced since I used ImageDataGenerator object, during training I used the training images as image generator and the training segments as mask generator and then passing them to the model. During testing, I have used the test images as image generator object, so the model is predicting the segments. That's what I gathered by reading up on how to perform segmentation using unet, and it seems to be working as well.
- Also, I tried to improve the UNet by increasing the number of epochs and the number of steps per epoch. I believe that should count as a valid improvement step.

**Comment by SiyuLiu0329** [comment]
> * The output is being produced since I used ImageDataGenerator object, during training I used the training images as image generator and the training segments as mask generator and then passing them to the model. During testing, I have used the test images as image generator object, so the model is predicting the segments. That's what I gathered by reading up on how to perform segmentation using unet, and it seems to be working as well.
> * Also, I tried to improve the UNet by increasing the number of epochs and the number of steps per epoch. I believe that should count as a valid improvement step.

I need to understand from a technical perspective, how you managed to convert sigmoid outputs (floats the range of 0-1) to integers?

**Comment by SiyuLiu0329** [comment]
Let me clearify. At best, your code would be okay for binary segmentation (1 class, sigmoid out + binaryCE loss). OASIS has 4 classes, you should be outputting 4 classes and use softmax CE loss.

**Comment by Calvin-Shaji** [comment]
To save the images, I used the img_as_ubyte function in the driver program to convert the output to unsigned 8-bit integer format.

**Comment by SiyuLiu0329** [comment]
> To save the images, I used the img_as_ubyte function in the driver program to convert the output to unsigned 8-bit integer format.

Can you let me know exactly how your model did 4-class segmentation given the current structure? I do not follow. It needs to predict 4 binary masks of 0s and 1s and be merged together using argmax. is your code doing the same thing. Otherwise it is not a  "valid" segmentation model and marks will be deducted.

**Comment by SiyuLiu0329** [comment]
There is also no one-hot encoding as the UNet requires

File: recognition/calvin/example_driver.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+import os
```
Comment (shakes76) [inline recognition/calvin/example_driver.py:1]: Code header block should be present on all scripts

**Comment by shakes76** [comment]
## TF Usage
OK
## Good Practice (Design/Commenting)
Good design
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage, Readme OK
Module present
Exact functionality of the code difficult to verify
## Commit Log
Meaningful commit messages 
Progressive commits used 
## Documentation
Description and Comments OK
Markdown used PDF submitted
## Pull Request
Partially Successful Pull Request (Working Algorithm Delivered on Time in Incorrect Branch) 
**Feedback required. Please choose the correct target branch topic-recognition**
**Remove unnecessary results image files, they will slow the Git repo down.**
No Request Description. In the future, a more detailed pull request could've been done. When developer receiving the request is not known, the request could be rejected. Remember you could use/reuse the README info.
